# Function to make the featurecounts.txt file and write it to this directory
make_featurecounts_file <- function(fc_nrow = 3000, filename = "featurecounts.txt", seed = 1) {
  if (file.exists('featurecounts.txt')) { return('featurecounts.txt exists, exiting the function') }
  # Create the same random data for every test
  set.seed(seed)
  # First row contains info on featureCounts command call including parameters and input files
  # First row of dataframe that will be exported as featurecounts.txt. Note that in featurecounts.txt
  # files that are generated by featurecounts only have 1 line for the first row.
  # in R, we need 12 columns in each row hence the 11 NAs added to the end of fc_row1
  fc_row1 <- c("# Program:featureCounts v2.0.3; Command:\"featureCounts\" \"-s2\" \"-p\" \"--countReadPairs\" \"-T\" \"16\" \"-a\" \"/data/NICHD-core0/references/mouse/gencode_m33/annotation/mouse_gencode_m33.gtf\" \"-o\" \"data/rnaseq_aggregation/featurecounts.txt\" \"data/rnaseq_samples/sample1/sample1.cutadapt.markdups.bam\" \"data/rnaseq_samples/sample2/sample2.cutadapt.markdups.bam\" \"data/rnaseq_samples/sample3/sample3.cutadapt.markdups.bam\" \"data/rnaseq_samples/sample4/sample4.cutadapt.markdups.bam\" \"data/rnaseq_samples/sample5/sample5.cutadapt.markdups.bam\" \"data/rnaseq_samples/sample6/sample6.cutadapt.markdups.bam\"", rep(NA, 11))


  # Create a data frame with example data
  data <- data.frame(
    Geneid = paste0("ENSMUSG", formatC(sample(10000:99999, fc_nrow, replace = FALSE), width = 10, flag = "0"),
                    ".", sample(1:20, fc_nrow, replace = TRUE)),  # Append version IDs to test strip dotted
    Chr = paste0("chr", sample(1:19, fc_nrow, replace = TRUE)),
    Start = sample(30000000:99990000, fc_nrow),
    End = sample(30001000:99999999, fc_nrow),
    Strand = sample(c("+", "-"), fc_nrow, replace = TRUE),
    Length = sample(500:15000, fc_nrow, replace = TRUE)
  )

  # Define sample paths based on lcfbwf featurecounts samplename format
  sample_names <- c(
    "data/rnaseq_samples/sample1/sample1.cutadapt.markdups.bam",
    "data/rnaseq_samples/sample2/sample2.cutadapt.markdups.bam",
    "data/rnaseq_samples/sample3/sample3.cutadapt.markdups.bam",
    "data/rnaseq_samples/sample4/sample4.cutadapt.markdups.bam",
    "data/rnaseq_samples/sample5/sample5.cutadapt.markdups.bam",
    "data/rnaseq_samples/sample6/sample6.cutadapt.markdups.bam"
  )
  # Simulate counts for each sample
  counts <- matrix(sample(0:20000, fc_nrow, replace = TRUE), ncol = 6)
  colnames(counts) <- sample_names
  # Combine gene data and counts
  feature_counts <- cbind(data, counts)
  fc_row2 <- colnames(feature_counts)
  colnames(feature_counts) <- NULL
  feature_counts <- rbind(fc_row2, feature_counts)
  # Add metadata to row1 saved in functions.R, colnames to row2 and data to the remaining `fc_nrow` rows
  feature_counts <-rbind(fc_row1, feature_counts)

  # Write the data frame to a text file
  write.table(feature_counts, file = filename, sep = "\t", quote = FALSE, row.names = FALSE)
  return(paste("File saved as", filename))
} # make_featureCounts_file

# Helper function to make minimal default design data. design_data is an argument and
# object of type list that is passed to make_dds()
make_design_data <- function() {
  make_featurecounts_file() # featurecounts.txt will be written to this directory if it doesn't exist
  lst <- list(
  # Create the sample table
  sampletable = data.frame(
    samplename = c("sample1", "sample2", "sample3", "sample4", "sample5", "sample6"),
    condition = c(rep("control", 3), rep("treatment", 3))),
  design = ~ condition
  ) # lst
  lst$sampletable$condition <- as.factor(lst$sampletable$condition)
  return(lst)
} # make_default_wald_design_data

make_dds_list <- function() {
  # Create design data and dds object for Wald test type
  wald_design_data <- make_design_data()
  make_featurecounts_file() # Write 'featurecounts.txt' if it does not exist
  dds_wald <- make_dds(wald_design_data,
                       config=config,
                       featureCounts='featurecounts.txt',
                       parallel=config$parallel$parallel)

  lrt_design_data <- make_lrt_design_data()
  dds_lrt <- make_dds(lrt_design_data,
                      config=config,
                      featureCounts='featurecounts.txt',
                      parallel=config$parallel$parallel)

  # Create dds_list
  dds_list <- list(dds_wald=dds_wald, dds_lrt=dds_lrt)
  return(dds_list)
} # make_dds_list

# Function to create design data for LRT test
make_lrt_design_data <- function() {
  lrt_design_data <- make_design_data()
  lrt_design_data$test <- 'LRT'
  lrt_design_data$reduced_design <- ~1
  return(lrt_design_data)
} # make_lrt_design_data

# Helper function to check the output of make_results
check_results <- function(res, lrt_design_data, label, contrast = NULL, coef = NULL, test = NULL, type = NULL) {
  print(label)
  # Check that results were returned by make_res
  expect_true(!is.null(res))
  expect_true(identical(names(res), c('res', 'dds', 'label')))
  # Check that the res element returned by make_res is a DESeqResults object
  expect_true(inherits(res$res, "DESeqResults"))
  # Save a character representing the source of LRT pvalue for comparison with make_results output
  lrt_mcols_description <- paste0(as.character(lrt_design_data$design)[1], " ",
                                  as.character(lrt_design_data$design)[2], "' vs '",
                                  as.character(lrt_design_data$reduced_design)[1], " ",
                                  as.character(lrt_design_data$reduced_design)[2], "'")

  # Check the metadata in res for correct test and coef/contrast
  # based on each combination of test, type and coef/contrast arguments

  if ((is.null(test) || test == 'Wald') && (!is.null(type) && type == 'ashr')) {
    # test == 'Wald' and type == 'ashr' -- OR -- test == NULL and type == 'ashr'
    expected_char <- paste(test %||% 'Wald', "test p-value:", contrast[1], contrast[2], "vs", contrast[3])
    expect_true(mcols(res$res)$description[4] == expected_char)
  } else if ((is.null(test) || test == 'Wald') && (!is.null(type) && type == 'apeglm')) {
    # test == 'Wald' and type == 'apeglm' -- OR -- test == NULL and type == 'apeglm'
    coef <- str_split(coef, "_")[[1]]
    expected_char <- paste(test %||% 'Wald', "test p-value:", coef[1], coef[2], coef[3], coef[4])
    expect_true(mcols(res$res)$description[4] == expected_char)
  } else if ((is.null(test) || test == 'Wald') && (is.null(type) || type == 'normal')) {
    # test == 'Wald' and type == 'normal', -- OR -- test == 'Wald' and type == NULL
    # test == NULL and type == 'normal' -- OR -- test == NULL and type == NULL
    expected_char <- paste(test %||% 'Wald', "statistic:", contrast[1], contrast[2], "vs", contrast[3])
    expect_true(mcols(res$res)$description[4] == expected_char)
  } else if ((!is.null(test) && test == 'LRT') && (!is.null(type) && type != 'normal')) {
    # test == 'LRT' and type == 'ashr' -- OR -- test == 'LRT and type == 'apeglm'
    expected_char <- paste0(test, " p-value: '", lrt_mcols_description)
    expect_true(mcols(res$res)$description[4] == expected_char)
  } else if ((!is.null(test) && test == 'LRT') && (is.null(type) || type == 'normal')) {
    # test == 'LRT and type == 'normal' -- OR -- test == 'LRT and type == NULL
    expected_char <- paste0(test, " statistic: '", lrt_mcols_description)
    expect_true(mcols(res$res)$description[4] == expected_char)
  } else {
    stop(paste(label, 'was not checked'))
  }

  # Check that the make_results defined metadata in the res object includes the correct shrinkage type
  if (!is.null(type)) {
    expect_true(identical(metadata(res$res)$type, type))
  } else if (is.null(type)) {
    expect_true(is.null(metadata(res$res)$type))
  }
} # check_results
